frontend/src/Frontend/Wallet.hs

```diff
@@ -129,78 +113,34 @@ makePactLenses ''WalletCfg
-    SomeAccount_Account a -> toJSON a
-
-instance FromJSON key => FromJSON (SomeAccount key) where
-  parseJSON Null = pure SomeAccount_Deleted
-  parseJSON x = (SomeAccount_Inflight <$> (withObject "Inflight Acc" (.: "inflight") x))
-    <|> SomeAccount_Account <$> parseJSON x
-
-activeAccountOnNetwork :: NetworkName -> SomeAccount key -> Maybe (Account key)
-activeAccountOnNetwork net = someAccount
-  Nothing
-  (\a -> a <$ guard (_account_network a == net))
-
-type Accounts key = IntMap (SomeAccount key)
-
-getSigningPairs :: Set AccountName -> Accounts key -> [KeyPair key]
-getSigningPairs signing = fmapMaybe isForSigning . IntMap.elems
+getSigningPairs :: KeyStorage key -> Accounts -> Set (Some AccountRef) -> [KeyPair key]
+getSigningPairs allKeys allAccounts signing = filterKeyPairs wantedKeys allKeys
   where
-    -- isJust filter is necessary so indices are guaranteed stable even after
-    -- the following `mapMaybe`:
-    isForSigning = \case
-      SomeAccount_Account a
-        | kp <- _account_key a
-        , Just _ <- _keyPair_privateKey kp
-        , Set.member (_account_name a) signing
-        -> Just kp
-      _ -> Nothing
+    wantedKeys = Set.fromList $ Map.elems $ Map.restrictKeys accountRefs signing
+    accountRefs = flip foldAccounts allAccounts $ \a@(r :=> _) -> Map.singleton (Some r) $ accountKey a
 
 data Wallet key t = Wallet
-  { _wallet_accounts :: Dynamic t (Accounts key)
+  { _wallet_keys :: Dynamic t (KeyStorage key)
+    -- ^ Accounts added and removed by the user
+  , _wallet_accounts :: Dynamic t AccountStorage
     -- ^ Accounts added and removed by the user
-  , _wallet_walletOnlyAccountCreated :: Event t AccountName
-    -- ^ A new wallet only account has been created
   }
   deriving Generic
 
 makePactLenses ''Wallet
 
-findFirstInflightAccount :: Accounts key -> Maybe (Account key)
-findFirstInflightAccount = fmap (\(SomeAccount_Inflight a) -> a) . find g
-  where g = \case SomeAccount_Inflight _ -> True; _ -> False
+findFirstInflightAccount :: Accounts -> Maybe (AccountName, ChainId, VanityAccount)
+findFirstInflightAccount as = do
+  (n, cm) <- Map.lookupMin $ _accounts_vanity as
+  (c, va) <- Map.lookupMin $ Map.filter _vanityAccount_inflight cm
+  pure (n, c, va)
 
 -- | Find the first vanity account in the wallet
-findFirstVanityAccount :: Accounts key -> Maybe (Account key)
-findFirstVanityAccount = fmap (\(SomeAccount_Account a) -> a) . find g
-  where g SomeAccount_Deleted = False
-        g (SomeAccount_Inflight _) = False
-        g (SomeAccount_Account a) = unAccountName (_account_name a) /= keyToText (_keyPair_publicKey $ _account_key a)
-
--- | An empty wallet that will never contain any keys.
-emptyWallet :: Reflex t => Wallet key t
-emptyWallet = Wallet mempty never
+findFirstVanityAccount :: Accounts -> Maybe (AccountName, ChainId, VanityAccount)
+findFirstVanityAccount as = do
+  (n, cm) <- Map.lookupMin $ _accounts_vanity as
+  (c, va) <- Map.lookupMin cm
+  pure (n, c, va)
```

Comment:
Seems like these are both used only once. How about instead defining only
```haskell
findFirstVanityAccount pred as = do
  (n, cm) <-  Map.lookupMin $ _accounts_vanity as
  (c, va) <- Map.lookupMin $ Map.filter pred cm
  pure (n, c, va)
```
and calling it as `findFirstVanityAccount _vanityAccount_inflight` or `findFirstVanityAccount (const True)`